libraries/core_libs/consensus/src/dag/sortition_params_manager.cpp

```diff
@@ -107,7 +113,28 @@ SortitionParams SortitionParamsManager::getSortitionParams(std::optional<uint64_
   return p;
 }
 
+uint16_t calculateEfficiencyHF2(const SyncBlock& block, uint16_t stale_difficulty) {
+  auto accLambda = [&stale_difficulty](uint32_t s, const DagBlock& b) -> uint32_t {
+    if (b.getDifficulty() == stale_difficulty) {
+      return s;
+    }
+    return s + b.getTrxs().size();
+  };
+  size_t total_count = std::accumulate(block.dag_blocks.begin(), block.dag_blocks.end(), 0, accLambda);
+  uint16_t res = 100 * kOnePercent;
+  if (total_count != 0) {
+    res = block.transactions.size() * 100 * kOnePercent / total_count;
```

Comment:
not 100% sure if thats what we really want if we want to ignore stale blocks and everything that is inside, I think you should iterate over the dag blocks in for loop, check if block is stale or not (as you do in lambda) and count only unique txs from non-stale blocks instead of `block.transactions.size()`

Reply:
Yes it seems like we exclude stale blocks only from the denominator, but not the numerator?   Which is why we have to check that we dont exceed 100%.   But why not exclude blocks from both denominator and numerator as @JakubFornadel  says?

Reply:
Yeah, that makes sense. But to check if that transaction is unique we should check it in the database, because transaction could be finalized in previous pbft blocks

Reply:
Now excluding from both, but not checking if transaction was finalised in previous pbft blocks

Reply:
hmm thats going to be pretty expensive but yeah I guess to be sure if tx is unique or not we need to also know if it wasnt included in some previous pbft block